CDC becomes unhealthy because of Reader with id MySqlReader has not received message for 158602 milliseconds"

This issue is obscured by #84 but if the CDC runs for any length of time it becomes unhealthy:

```
{
  "status": "DOWN",
  "details": {
    "binlogEntryReaderHealthCheck": {
      "status": "DOWN",
      "details": {
        "error-1": "Reader with id MySqlReader has not received message for 266718 milliseconds",
        "detail-1": "Reader with id MySqlReader is connected"
      }
    },
    "cdcDataPublisherHealthCheck": {
      "status": "UP"
    },
    "zookeeperHealthCheck": {
      "status": "UP",
      "details": {
        "detail-1": "Connected to Zookeeper"
      }
    },
    "kafkaHealthCheck": {
      "status": "UP",
      "details": {
        "detail-1": "Connected to Kafka"
      }
    },
    "diskSpace": {
      "status": "UP",
      "details": {
        "total": 62725623808,
        "free": 33846505472,
        "threshold": 10485760
      }
    }
  }
}

```

Investigated.

log:

0) used zero for last event time (https://github.com/eventuate-foundation/eventuate-cdc/issues/84), cdc starts and immediately reports that it is unhealthy. 
1) Found that monitoring inserted and updated.
2) Found that update rows event is not received.
3) Debugged shyiko connector.
Found following unlogged exception:

exception `com.github.shyiko.mysql.binlog.event.deserialization.EventDataDeserializationException: Failed to deserialize data of EventHeaderV4{timestamp=1605779240000, eventType=EXT_WRITE_ROWS, serverId=1, headerLength=19, dataLength=22, nextPosition=1049, flags=0}`


cause `com.github.shyiko.mysql.binlog.event.deserialization.MissingTableMapEventException: No TableMapEventData has been found for table id:108. Usually that means that you have started reading binary log 'within the logical event group' (e.g. from WRITE_ROWS and not proceeding TABLE_MAP`

4) checked related issues on shyiko connector github page. But did not found a clue. Continued debugging.

5) found that connector expects TableMapEventData instance in `Map<Long, TableMapEventData> tableMapEventByTableId`

in update deserializer.

6) I found that by mistake I deleted adding to tableMapEventByTableId monitoring table map event:

https://github.com/eventuate-foundation/eventuate-cdc/commit/bcdf3cd970fb11fe2b688df44595950113f97be9#diff-450e9de12d2111eb2f2b87f240350cf1867ea1ae6411aae280fecac9d0187787R221

because cdc monitoring change does not require any special handling like here: https://github.com/dartartem/eventuate-cdc-1/blob/wip-db-id-gen/eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java#L303

7) Fixed. Now cdc is reported healthy. Because issue is reproducible by existing tests, I did not add new. Please see point 0.

8) Added [lifecycle handler](https://github.com/dartartem/eventuate-cdc-1/blob/cb3bd8315f8acd277edb6684ab782463fadaddbd/eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClientLifecycleListener.java) to show exceptions like from point 3
But now on cdc start following exceptions appeared:
```

com.github.shyiko.mysql.binlog.event.deserialization.EventDataDeserializationException: Failed to deserialize data of EventHeaderV4{timestamp=1605800408000, eventType=EXT_UPDATE_ROWS, serverId=1, headerLength=19, dataLength=61, nextPosition=5506, flags=0}
	at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:233) ~[mysql-binlog-connector-java-0.16.1.jar:0.16.1]
	at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.nextEvent(EventDeserializer.java:201) ~[mysql-binlog-connector-java-0.16.1.jar:0.16.1]
	at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:890) [mysql-binlog-connector-java-0.16.1.jar:0.16.1]
	at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:559) [mysql-binlog-connector-java-0.16.1.jar:0.16.1]
	at com.github.shyiko.mysql.binlog.BinaryLogClient$7.run(BinaryLogClient.java:793) [mysql-binlog-connector-java-0.16.1.jar:0.16.1]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_131]
Caused by: com.github.shyiko.mysql.binlog.event.deserialization.MissingTableMapEventException: No TableMapEventData has been found for table id:108. Usually that means that you have started reading binary log 'within the logical event group' (e.g. from WRITE_ROWS and not proceeding TABLE_MAP
	at com.github.shyiko.mysql.binlog.event.deserialization.AbstractRowsEventDataDeserializer.deserializeRow(AbstractRowsEventDataDeserializer.java:104) ~[mysql-binlog-connector-java-0.16.1.jar:0.16.1]
	at io.eventuate.local.mysql.binlog.UpdateRowsDeserializer.deserializeRow(UpdateRowsDeserializer.java:23) ~[main/:na]
	at com.github.shyiko.mysql.binlog.event.deserialization.UpdateRowsEventDataDeserializer.deserializeRows(UpdateRowsEventDataDeserializer.java:71) ~[mysql-binlog-connector-java-0.16.1.jar:0.16.1]
	at com.github.shyiko.mysql.binlog.event.deserialization.UpdateRowsEventDataDeserializer.deserialize(UpdateRowsEventDataDeserializer.java:58) ~[mysql-binlog-connector-java-0.16.1.jar:0.16.1]
	at com.github.shyiko.mysql.binlog.event.deserialization.UpdateRowsEventDataDeserializer.deserialize(UpdateRowsEventDataDeserializer.java:33) ~[mysql-binlog-connector-java-0.16.1.jar:0.16.1]
	at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:227) ~[mysql-binlog-connector-java-0.16.1.jar:0.16.1]
	... 5 common frames omitted
```

Appeared on cdc start on clean database.
Not yet investigated. Probably some database internal operations different from message/monitoring writes.
**Probably it is worth to add all TABLE_MAP events to** `Map<Long, TableMapEventData> tableMapEventByTableId` in [TableMapper](https://github.com/eventuate-foundation/eventuate-cdc/blob/wip-db-id-gen/eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/TableMapper.java).

--- a/eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/BinlogEntryReader.java
+++ b/eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/BinlogEntryReader.java
@@ -28,7 +28,7 @@ public abstract class BinlogEntryReader {
   protected CommonCdcMetrics commonCdcMetrics;
   protected volatile Optional<String> processingError = Optional.empty();
 
-  private volatile long lastEventTime = System.currentTimeMillis();
+  private volatile long lastEventTime = 0;
 
   protected Optional<Runnable> restartCallback = Optional.empty();
 

--- a/eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java
+++ b/eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java
@@ -165,6 +165,8 @@ public class MySqlBinaryLogClient extends DbLogClient {
 
     client.registerEventListener(eventListener);
 
+    client.registerLifecycleListener(new MySqlBinaryLogClientLifecycleListener(readerName));
+
     connectWithRetriesOnFail();
 
     try {
@@ -204,6 +206,7 @@ public class MySqlBinaryLogClient extends DbLogClient {
         TableMapEventData tableMapEvent = event.getData();
 
         if (cdcMonitoringDao.isMonitoringTableChange(tableMapEvent.getDatabase(), tableMapEvent.getTable())) {
+          tableMapper.addMapping(tableMapEvent);
           cdcMonitoringTableId = Optional.of(tableMapEvent.getTableId());
           break;
         }
@@ -218,7 +221,7 @@ public class MySqlBinaryLogClient extends DbLogClient {
                 .anyMatch(schemaAndTable::equals);
 
         if (shouldHandleTable) {
-          if (tableMapper.addMapping(tableMapEvent)) {
+          if (tableMapper.addMappingAndCheckIfColumnRefreshIsNecessary(tableMapEvent)) {
            mySqlBinlogEntryExtractor.refreshColumnOrder();
           }
         } else {

--- a/eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClientLifecycleListener.java
+++ b/eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClientLifecycleListener.java
@@ -1,0 +1,37 @@
+package io.eventuate.local.mysql.binlog;
+
+import com.github.shyiko.mysql.binlog.BinaryLogClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MySqlBinaryLogClientLifecycleListener implements BinaryLogClient.LifecycleListener {
+  private Logger logger = LoggerFactory.getLogger(getClass());
+
+  private String readerName;
+
+  public MySqlBinaryLogClientLifecycleListener(String readerName) {
+    this.readerName = readerName;
+  }
+
+  @Override
+  public void onConnect(BinaryLogClient client) {
+    logger.info("Reader {} connected.", readerName);
+  }
+
+  @Override
+  public void onCommunicationFailure(BinaryLogClient client, Exception ex) {
+    logger.error("Reader {} communication failed.", readerName);
+    logger.error("Reader communication failed.", ex);
+  }
+
+  @Override
+  public void onEventDeserializationFailure(BinaryLogClient client, Exception ex) {
+    logger.error("Reader {} deserialization failed.", readerName);
+    logger.error("Reader deserialization failed.", ex);
+  }
+
+  @Override
+  public void onDisconnect(BinaryLogClient client) {
+    logger.info("Reader {} disconnected.", readerName);
+  }
+}

--- a/eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/TableMapper.java
+++ b/eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/TableMapper.java
@@ -21,12 +21,16 @@ public class TableMapper {
     return new SchemaAndTable(tableMapEventData.getDatabase(), tableMapEventData.getTable());
   }
 
-  public boolean addMapping(TableMapEventData tableMapEventData) {
+  public boolean addMappingAndCheckIfColumnRefreshIsNecessary(TableMapEventData tableMapEventData) {
     boolean refresh = shouldRefreshColumns(tableMapEventData);
 
-    tableMapEventByTableId.put(tableMapEventData.getTableId(), tableMapEventData);
+    addMapping(tableMapEventData);
 
     return refresh;
+  }
+
+  public void addMapping(TableMapEventData tableMapEventData) {
+    tableMapEventByTableId.put(tableMapEventData.getTableId(), tableMapEventData);
   }
 
   public boolean isTableMapped(long tableId) {

